Add script for ignored files in the working tree

pluehne opened the pull request:
This adds a script that lists all files that are ignored by `.gitignore` but present in the worktree nevertheless. The option `--sort-by-size` sorts the output by file size, whereas the default is to sort the files lexicographically with numeric sorting.

pluehne committed:
Add script for ignored files in the working tree

This adds a script listing all files that are ignored by .gitignore but
present in the working tree nevertheless. The option “--sort-by-size”
sorts the output by file size, whereas the default is to sort the files
lexicographically with numeric sorting.

pluehne commented:
@mlbright: For your interest :slightly_smiling_face:.

larsxschneider commented on git-find-ignored-files.sh:
```diff
@@ -0,0 +1,43 @@
+		exit 0
+		;;
+	-s|--sort-by-size)
+		find . -type f -print0 | git check-ignore --no-index -z --stdin | xargs -0 du -shc | sort -h
```
nit-pick: Do you see a way to make that a bit less redundant? E.g. you could define a variable here and then `eval` the stuff below?

pluehne replied:
Will do!

larsxschneider commented on git-find-ignored-files.sh:
```diff
@@ -0,0 +1,43 @@
+
+function print_help
+{
+	echo "usage: git-find-ignored-files [-s | --sort-by-size] [--help]"
```
Here is a trick how you could write the help in the script header and print it:
https://github.com/larsxschneider/cached-clone/blob/c2b7ca670da7854e7139e40ce8363cdad3874e8d/git-cclone#L38-L39

pluehne replied:
Thanks for the pointer, I’ll adjust the usage output :+1:.

larsxschneider reviewed (changes requested):
Very nice idea 👍

pluehne commented:
@larsxschneider: I think I foud a solution that is cleaner in multiple respects:

1. Instead of `find . -type f -print0 | git check-ignore --no-index -z --stdin`, my new solution simply uses `git ls-files --ignored --exclude-standard -z`, which also does the trick but is much more concise
1. I assemble the command in a Bash variable and execute the final statement with `eval`
1. By adding the `-r` option to `xargs`, this prevents unwanted output in the case that no ignored files are present in the working tree or index.
1. Finally, I also adjusted printing the way the usage information is obtained.

I’d be happy if you could review my changes :smiley: again.

larsxschneider commented on git-find-ignored-files.sh:
```diff
@@ -0,0 +1,43 @@
+	exit 1
+fi
+
+command="git ls-files --ignored --exclude-standard -z | xargs -0r du -sh"
```
`-r` seems not to be available on the macOS version xargs version:
```
./git-find-ignored-files.sh
xargs: illegal option -- r
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
```

pluehne committed:
Disable xargs’s -r option on macOS

This implements a switch that doesn’t set the -r option in xargs if it
isn’t supported. In most xargs implementations without a -r option, the
default is not to run the command with empty input anyway, making this
option unnecessary.

pluehne commented:
@larsxschneider: I now had time to come back to this pull request now. I’ve first tried to work around the missing `-r` option on macOS by storing the results of the `git ls-files` call in a shell variable. However, this turned out to be difficult, because the null bytes aren’t properly piped to `xargs` when `echo`ing or `printf`ing the contents of the variable.

Instead, I now implemented a switch that detects `xargs`’s feature support for `-r`. This is a bit ugly, but it should do the trick. Note that on macOS, specifying the `-r` option is redundant, because commands in macOS’s implementation of `xargs` aren’t run when the `stdin` is empty. That’s not the case for many other implementations, though, for which my workaround would still specify `-r` if it exists.

larsxschneider commented:
Nice, that works! 👍 

One more nit tough: the original script in our product-migrations-repo prints the responsible `.gitignore` line for each ignore. That helps to find the culprit. Can you add that too?

pluehne committed:
Add gitignore origin for each file to output

In addition to the file name and size, this commit adds a third column
to the output that, for each file, shows why it is ignored.

In order to accomplish this, the code was restructured to feed the list
of ignored files to two separate commands for 1) computing the file size
and 2) obtaining the gitignore origin.

Due to this refactoring, the previous xargs-related switch for macOS
systems can be safely removed, because the file list can now be checked
for emptiness before running the subsequent commands (instead of having
to resort to the “-r” option of xargs).

pluehne commented:
@larsxschneider: As you requested, I added a third column to the output that contains the gitignore origin. The implementation was slightly tricky, because I needed to run both `du -sh` and `git check-ignore` on the same input. That’s why I needed to introduce temporary variables storing the results etc. (find more detailed explanations in my commit message in 4fdb798bde40d86145e7dff45679bae495745d2e). The upside is that I could get rid of the quirk to support `xargs` on macOS from the prior commit.

larsxschneider reviewed (approved):
Looks great! Thank you!
